### PR TITLE
Adding version 0.4.7 of simdjson (up from just having version 0.2.1)

### PR DESCRIPTION
**Pull request by lemire**
Specify library name and version:  simdjson/0.4.7

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

This PR follows up on https://github.com/conan-io/conan-center-index/pull/2104

cc @uilianries @prince-chrismc

### CODE REVIEW

**Comment by prince-chrismc** [comment]
:boom: that one white space :rofl:

**Comment by lemire** [comment]
@prince-chrismc It is the same files brought over from the other PR but I wanted to bring in your changes, so I copied and pasted. My editor (vim) is not great at copy-paste as it will often introduce spurious spacing.

No matter. I will get it right eventually.

**Comment by lemire** [comment]
The error before my last commit (based on previously proposed changes) was...

```
simdjson/0.4.7: WARN: Build folder is dirty, removing it: /home/conan/workspace/onan-center-pull-request_PR-2249/4/0_19/.conan/data/simdjson/0.4.7/_/_/build/f40d55787723a498ce174707c707fe5cfcf8e46f
simdjson/0.4.7: WARN: Trying to remove corrupted source folder
simdjson/0.4.7: WARN: This can take a while for big packages
ERROR: simdjson/0.4.7: Error in source() method, line 64
	strict=False)
	FileNotFoundError: [Errno 2] No such file or directory: 'source_subfolder/tools/cmake/FindOptions.cmake'
```

**Comment by uilianries** [comment]
Is not possible sharing the same Conanfile for both versions? Separated recipes increase the maintenance level

**Comment by lemire** [comment]
> Is not possible sharing the same Conanfile for both versions? Separated recipes increase the maintenance level

I don't think that's the problem. The real problem is that you apparently can't test these builds yourself. You have to rely on this ***super*** slow CI system.

When I run local tests, things apparently work. When I push and wait an hour, I find out that it does not work.

It is very inefficient.

That's the real problem.

If I could debug it locally, that would be much, much, much better.

**Comment by prince-chrismc** [comment]
You can run `conan create {folder} {version}@` In the wiki there's a section for setting up the hooks. You should be able to do 90% of the testing locally. 

How are you trying to teat locally?

**Comment by lemire** [comment]
> You can run conan create {folder} {version}@ In the wiki there's a section for setting up the hooks. You should be able to do 90% of the testing locally. How are you trying to teat locally?

Yes. And it works fine.

**Comment by lemire** [comment]
These are not the kind of errors that I expect to encounter...

```

/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/package/db17f6328d44f601d07cb1d228fe2abc89aa2349/lib/libsimdjson.a(simdjson.cpp.o): in function `__cxx_global_var_init':
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/iostream:74: undefined reference to `std::ios_base::Init::Init()'
/usr/bin/ld: /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/iostream:74: undefined reference to `std::ios_base::Init::~Init()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/package/db17f6328d44f601d07cb1d228fe2abc89aa2349/lib/libsimdjson.a(simdjson.cpp.o): in function `__cxx_global_var_init.1':
/home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:7: undefined reference to `std::allocator<char>::allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:8: undefined reference to `std::allocator<char>::allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:9: undefined reference to `std::allocator<char>::allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:10: undefined reference to `std::allocator<char>::allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:11: undefined reference to `std::allocator<char>::allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/package/db17f6328d44f601d07cb1d228fe2abc89aa2349/lib/libsimdjson.a(simdjson.cpp.o):/home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:12: more undefined references to `std::allocator<char>::allocator()' follow
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/package/db17f6328d44f601d07cb1d228fe2abc89aa2349/lib/libsimdjson.a(simdjson.cpp.o): in function `__cxx_global_var_init.1':
/home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:6: undefined reference to `std::allocator<char>::~allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:6: undefined reference to `std::allocator<char>::~allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:6: undefined reference to `std::allocator<char>::~allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:6: undefined reference to `std::allocator<char>::~allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:6: undefined reference to `std::allocator<char>::~allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/package/db17f6328d44f601d07cb1d228fe2abc89aa2349/lib/libsimdjson.a(simdjson.cpp.o):/home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/src/error.cpp:6: more undefined references to `std::allocator<char>::~allocator()' follow
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/package/db17f6328d44f601d07cb1d228fe2abc89aa2349/lib/libsimdjson.a(simdjson.cpp.o): in function `implementation':
/home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/include/simdjson/implementation.h:135: undefined reference to `std::allocator<char>::allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/include/simdjson/implementation.h:135: undefined reference to `std::allocator<char>::~allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/include/simdjson/implementation.h:136: undefined reference to `std::allocator<char>::allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/include/simdjson/implementation.h:135: undefined reference to `std::allocator<char>::~allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/include/simdjson/implementation.h:136: undefined reference to `std::allocator<char>::~allocator()'
/usr/bin/ld: /home/conan/workspace/onan-center-pull-request_PR-2249/7/0_39/.conan/data/simdjson/0.4.7/_/_/build/db17f6328d44f601d07cb1d228fe2abc89aa2349/source_subfolder/include/simdjson/implementation.h:139: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'

```

**Comment by lemire** [comment]
Locally, things are fine.

Screen Shot 2020-07-21 at 4 13 24 PM

**Comment by lemire** [comment]
I have linkage problems...

```
[100%] Linking CXX executable bin/test_package
/home/conan/workspace/onan-center-pull-request_PR-2249/7/0_43/.conan/data/simdjson/0.4.7/_/_/package/ac1063b7ee62e8b348d858ecd47a5adbf6fc2fd7/lib/libsimdjson.a(simdjson.cpp.o): In function `_GLOBAL__sub_I_simdjson.cpp':
simdjson.cpp:(.text.startup+0x250): undefined reference to `std::ios_base::Init::Init()'
simdjson.cpp:(.text.startup+0x257): undefined reference to `std::ios_base::Init::~Init()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
CMakeFiles/test_package.dir/build.make:83: recipe for target 'bin/test_package' failed
make[2]: *** [bin/test_package] Error 1
CMakeFiles/Makefile2:75: recipe for target 'CMakeFiles/test_package.dir/all' failed
make[1]: *** [CMakeFiles/test_package.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by lemire** [comment]
@prince-chrismc  @uilianries 

Ok. Without access to the testing environment, I am not going to be able to debug it. It just too slow.

Here is our CI:

Github action: https://github.com/simdjson/simdjson/tree/master/.github/workflows
Drone: https://github.com/simdjson/simdjson/blob/master/.drone.yml
Circle CI: https://github.com/simdjson/simdjson/blob/master/.circleci/config.yml

We have simdjson bindings deployed for 0.4.x in Python and several other systems. 

This code should work. It should build just fine.

There are strange warnings...

```
  Manually-specified variables were not used by the project:

    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    CONAN_CMAKE_POSITION_INDEPENDENT_CODE
    CONAN_COMPILER
    CONAN_COMPILER_VERSION
    CONAN_CXX_FLAGS
    CONAN_C_FLAGS
    CONAN_EXPORTED
    CONAN_IN_LOCAL_CACHE
    CONAN_LIBCXX
    CONAN_SHARED_LINKER_FLAGS
```

Is our CMake supposed to use those?

**Comment by lemire** [comment]
Now with Visual Studio....

```

simdjson.lib(simdjson.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value 'MTd_StaticDebug' in test_package.obj [C:\J\workspace\onan-center-pull-request_PR-2249\9\co_0_24\recipes\simdjson\0.4.7\test_package\build\3294438b7a141da88b6008adf3bb53d3c0bb43e4\test_package.vcxproj]

msvcprtd.lib(MSVCP140D.dll) : error LNK2005: "public: __cdecl std::_Lockit::_Lockit(int)" (??0_Lockit@std@@QEAA@H@Z) already defined in libcpmtd.lib(xlock.obj) [C:\J\workspace\onan-center-pull-request_PR-2249\9\co_0_24\recipes\simdjson\0.4.7\test_package\build\3294438b7a141da88b6008adf3bb53d3c0bb43e4\test_package.vcxproj]

msvcprtd.lib(MSVCP140D.dll) : error LNK2005: "public: __cdecl std::_Lockit::~_Lockit(void)" (??1_Lockit@std@@QEAA@XZ) already defined in libcpmtd.lib(xlock.obj) [C:\J\workspace\onan-center-pull-request_PR-2249\9\co_0_24\recipes\simdjson\0.4.7\test_package\build\3294438b7a141da88b6008adf3bb53d3c0bb43e4\test_package.vcxproj]

msvcprtd.lib(MSVCP140D.dll) : error LNK2005: "public: struct _Cvtvec __cdecl std::_Locinfo::_Getcvt(void)const " (?_Getcvt@_Locinfo@std@@QEBA?AU_Cvtvec@@XZ) already defined in libcpmtd.lib(locale.obj) [C:\J\workspace\onan-center-pull-request_PR-2249\9\co_0_24\recipes\simdjson\0.4.7\test_package\build\3294438b7a141da88b6008adf3bb53d3c0bb43e4\test_package.vcxproj]

msvcprtd.lib(MSVCP140D.dll) : error LNK2005: "public: char const * __cdecl std::_Locinfo::_Getdays(void)const " (?_Getdays@_Locinfo@std@@QEBAPEBDXZ) already defined in libcpmtd.lib(wlocale.obj) [C:\J\workspace\onan-center-pull-request_PR-2249\9\co_0_24\recipes\simdjson\0.4.7\test_package\build\3294438b7a141da88b6008adf3bb53d3c0bb43e4\test_package.vcxproj]

msvcprtd.lib(MSVCP140D.dll) : error LNK2005: "public: char const * __cdecl std::_Locinfo::_Getmonths(void)const " (?_Getmonths@_Locinfo@std@@QEBAPEBDXZ) already defined in libcpmtd.lib(wlocale.obj) [C:\J\workspace\onan-center-pull-request_PR-2249\9\co_0_24\recipes\simdjson\0.4.7\test_package\build\3294438b7a141da88b6008adf3bb53d3c0bb43e4\test_package.vcxproj]

msvcprtd.lib(MSVCP140D.dll) : error LNK2005: "public: unsigned short const * __cdecl std::_Locinfo::_W_Getdays(void)const " (?_W_Getdays@_Locinfo@std@@QEBAPEBGXZ) already defined in libcpmtd.lib(wlocale.obj) [C:\J\workspace\onan-center-pull-request_PR-2249\9\co_0_24\recipes\simdjson\0.4.7\test_package\build\3294438b7a141da88b6008adf3bb53d3c0bb43e4\test_package.vcxproj]

```

**Comment by uilianries** [comment]
Hi @lemire ! Thanks again for your contribution, please take a look: https://github.com/lemire/conan-center-index/pull/1. That PR should fix the linux error, Regards!

**Comment by lemire** [comment]
After getting some help from @uilianries, I have tried to re-enable more compilers.

File: recipes/simdjson/0.4.7/test_package/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.9)
```
Comment (uilianries) [inline recipes/simdjson/0.4.7/test_package/CMakeLists.txt:1]:
```suggestion
cmake_minimum_required(VERSION 3.1)
```

Cmake 3.1 is enough here.

**Comment by lemire** [comment]
@uilianries The remaining issue is with visual studio....

```


test_package.obj : error LNK2001: unresolved external symbol "class simdjson::internal::atomic_ptr<class simdjson::implementation const > simdjson::active_implementation" (?active_implementation@simdjson@@3V?$atomic_ptr@$$CBVimplementation@simdjson@@@internal@1@A) [C:\J\workspace\onan-center-pull-request_PR-2249\12\co_0_57\recipes\simdjson\0.4.7\test_package\build\cb20919d00fb124356a96568f7b0438954f6c8ec\test_package.vcxproj]

test_package.obj : error LNK2001: unresolved external symbol "struct simdjson::internal::error_code_info const * const simdjson::internal::error_codes" (?error_codes@internal@simdjson@@3QBUerror_code_info@12@B) [C:\J\workspace\onan-center-pull-request_PR-2249\12\co_0_57\recipes\simdjson\0.4.7\test_package\build\cb20919d00fb124356a96568f7b0438954f6c8ec\test_package.vcxproj]

C:\J\workspace\onan-center-pull-request_PR-2249\12\co_0_57\recipes\simdjson\0.4.7\test_package\build\cb20919d00fb124356a96568f7b0438954f6c8ec\bin\test_package.exe : fatal error LNK1120: 2 unresolved externals [C:\J\workspace\onan-center-pull-request_PR-2249\12\co_0_57\recipes\simdjson\0.4.7\test_package\build\cb20919d00fb124356a96568f7b0438954f6c8ec\test_package.vcxproj]



```

It looks like it is not exporting symbols.

Can you help?

**Comment by uilianries** [comment]
@lemire if simdjson can not be used as shared library on Windows, you can drop the support:

```python
def configure(self):
    if self.settings.os == "Windows" and self.options.shared:
        raise ConanInvalidConfiguration("simdjson can be build as shared library in Windows.")
```

**Comment by uilianries** [comment]
Okay, I gonna connect to my Windows machine and check it

**Comment by lemire** [comment]
>  if simdjson can not be used as shared library on Windows, you can drop the support:

But it can be!

**Comment by lemire** [comment]
This works just fine under Visual Studio...

```
mkdir build
cd build
cmake -DSIMDJSON_BUILD_STATIC=OFF ..
cmake --build . --target basictests --config Release
ctest . -R basic
```

File: recipes/simdjson/0.4.7/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        self.copy("LICENSE", src=self._source_subfolder, dst="licenses")
+
+    def package_info(self):
+        self.cpp_info.libs = ['simdjson']
+        if self.settings.os == "Linux":
+            self.cpp_info.system_libs = ["m"]
+        if self.options.threads:
+            self.cpp_info.defines = ["SIMDJSON_THREADS_ENABLED=1"]
+            if self.settings.os == "Linux":
+                self.cpp_info.system_libs.append("pthread")
```
Comment (uilianries) [inline recipes/simdjson/0.4.7/conanfile.py:96]:
```suggestion
                self.cpp_info.system_libs.append("pthread")
            if self.options.shared:
                self.cpp_info.defines.append("SIMDJSON_USING_LIBRARY=1")
```

Forgot to expose it ... that's the windows shared fix.

File: recipes/simdjson/0.4.7/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 2.8)
+project(cmake_wrapper)
+
+include("conanbuildinfo.cmake")
+conan_basic_setup()
+if(WIN32 AND BUILD_SHARED_LIBS)
+    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
+endif()
```
Comment (uilianries) [inline recipes/simdjson/0.4.7/CMakeLists.txt:8]:
```suggestion

```

I don't think you need to force it as simdjson already exposed it: https://github.com/simdjson/simdjson/blob/v0.4.7/src/CMakeLists.txt#L41

**Comment by uilianries** [comment]
@lemire All good !!:tada:

**Comment by lemire** [comment]
@uilianries  Yes. Thanks. This does look good now.  It is very clean.

**Pull request merged**